server/engine/src/main/java/org/eclipse/lsp/cobol/dialects/ibm/PreprocessorStage.java

```diff
@@ -88,4 +106,77 @@ private CopybooksRepository runPreprocessor(String programDocumentUri, AnalysisC
     ctx.getAccumulatedErrors().addAll(preprocessorErrors);
     return copybooks;
   }
+
+  private void processCobolJavaInteroperabilityDirectives(AnalysisContext ctx) {
+    String text = ctx.getExtendedDocument().getCurrentText().toString();
+    Pattern compilerDirectiveLine =
+        Pattern.compile("(?i)(?:\\d.{5}.*|\\s*)>>\\s*(?<compilerDirectives>.+)");
+    Pattern newLinePattern = Pattern.compile("\n\r?");
+    Pattern sectionPattern =
+        Pattern.compile(
+            "(?i)\\s*DATA\\s+DIVISION.*|\\s*WORKING-STORAGE.*|\\s*PROCEDURE\\s+DIVISION.*");
+    Pattern javaShareableonPattern = Pattern.compile("(?i)\\s*>>\\s?JAVA-SHAREABLE\\s+ON\\s*");
+
+    String[] lines = newLinePattern.split(text);
+    String section = "";
+    boolean isJavaShareableOn = false;
+    for (int i = 0; i < lines.length; i++) {
+      Matcher directivesLine = compilerDirectiveLine.matcher(lines[i]);
+      Matcher sectionLine = sectionPattern.matcher(lines[i]);
+      if (!isJavaShareableOn && javaShareableonPattern.matcher(lines[i]).matches()) {
+        isJavaShareableOn = true;
+      }
+      if (sectionLine.matches()) {
+        section = sectionLine.group().trim();
+      }
+      if (!directivesLine.matches()) {
+        continue;
+      }
+
+      String compilerDirectives = directivesLine.group("compilerDirectives");
+      if (compilerDirectives != null) {
+        ctx.getDialectNodes()
+            .addAll(
+                process(
+                    compilerDirectives,
+                    ctx,
+                    new Position(i, directivesLine.start("compilerDirectives")),
+                    section,
+                    directivesLine.group(),
+                    isJavaShareableOn));
```

Comment:
All of this has already been parsed with the various regexes, why are you initializing a new lexer, parser and error handler for individual lines?